Have RenderObject subclass CanMakeCheckedPtr

cdumez opened the pull request:
#### 60da0af827823f29aa62dd74f9dc71deafab48f8
<pre>
Have RenderObject subclass CanMakeCheckedPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=261902">https://bugs.webkit.org/show_bug.cgi?id=261902</a>

Reviewed by Simon Fraser.

Have RenderObject subclass CanMakeCheckedPtr so that we can adopt CheckedPtr / CheckedRef
in rendering code for safety. We were previously relying on WeakPtr but it turned out to
be too costly and cause performance regressions.

This tested as perf-neutral on Speedometer, Membuster and MotionMark.

In order to do this without regressing Membuster, I had to re-order some data members and
rely on PackedPtr&lt;&gt; to not increase the size of renderers.

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::RenderElement):
(WebCore::RenderElement::attachRendererInternal):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::firstChild const):
(WebCore::RenderElement::lastChild const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: <a href="https://commits.webkit.org/268298@main">https://commits.webkit.org/268298@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3895802849993bee9db0acb28eb279b89bcc7eac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19660 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21134 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18003 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19786 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19666 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19463 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19525 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16733 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22002 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16713 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17520 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23866 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17773 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17695 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21827 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18277 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15478 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17412 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4607 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21769 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18111 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -110,18 +110,22 @@ RenderObject::SetLayoutNeededForbiddenScope::~SetLayoutNeededForbiddenScope()
 #endif
 
-struct SameSizeAsRenderObject : CanMakeWeakPtr<SameSizeAsRenderObject> {
+struct SameSizeAsRenderObject : WeakPtr<SameSizeAsRenderObject> {
```
Why isn't this using `CanMakeCheckedPtr` ?

cdumez replied:
I didn't mean to make this change, should remain `CanMakeWeakPtr< SameSizeAsRenderObject >`.

The issue we have the following data member below:
```
WeakHashSet<void*> cachedResourceClientAssociatedResources;
```

It comes from CachedImageClient, which is what RenderObject actually subclasses. But for some reason, we decided to have SameSizeAsRenderObject subclass `WeakPtr<SameSizeAsRenderObject>` instead of `CachedImageClient`. Instead we subclass `CanMakeWeakPtr` which `CachedImageClient` inherits and then include the CachedImageClient data member below. Not sure what's preferred. Let me know.

cdumez replied:
So we have 2 possibilities, either:
1. We have `SameSizeAsRenderObject` subclass `CachedImageClient` and `CanMakeCheckedPtr`, like `RenderObject` does. And then remove the `cachedResourceClientAssociatedResources` member from `SameSizeAsRenderObject`.
2. Do what I did in this patch, keep subclassing `CanMakeWeakPtr` and add the data members inherited from `CanMakeCheckedPtr`, similarly to what was done with `cachedResourceClientAssociatedResources`.

Let me know which you prefer.

cdumez committed:
Have RenderObject subclass CanMakeCheckedPtr
https://bugs.webkit.org/show_bug.cgi?id=261902

Reviewed by Simon Fraser.

Have RenderObject subclass CanMakeCheckedPtr so that we can adopt CheckedPtr / CheckedRef
in rendering code for safety. We were previously relying on WeakPtr but it turned out to
be too costly and cause performance regressions.

This tested as perf-neutral on Speedometer, Membuster and MotionMark.

In order to do this without regressing Membuster, I had to re-order some data members and
rely on PackedPtr<> to not increase the size of renderers.

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::RenderElement):
(WebCore::RenderElement::attachRendererInternal):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::firstChild const):
(WebCore::RenderElement::lastChild const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: https://commits.webkit.org/268298@main
